### PR TITLE
refactor ma_convproc_tend (part 1)

### PR DESCRIPTION
**Pull request by tangshuaiqi**
The subroutine ma_convproc_tend includes three layers of loops:
1. loop each column
         2. loop each time substep
                 3. loop each vertical level (from cloud base to cloud top)

This PR refactor part of the subroutine ma_convproc_tend:
- refactor all loops in the 3rd layer (from cloud base to cloud top) and separate into subroutines
- clean up some variables (fracice, dconudt_aqchem)
- change dp_i to dpdry_i and define new dp_i as dp(icol, :)

### CODE REVIEW

**Review by overfelt — APPROVED** [review]

**Review by jeff-cohere — APPROVED** [review]

File: components/eam/src/chemistry/modal_aero/modal_aero_convproc.F90 [hunk 1]
```diff
@@ -1786,12 +1468,556 @@ subroutine ma_convproc_tend(                                           &
...
+   real(r8),intent(in)  :: dpdry_i(pver)        ! dp [mb]
+   real(r8),intent(in)  :: cldfrac(pcols, pver) ! cloud fraction [fraction]
+   real(r8),intent(in)  :: rhoair_i(pver)       ! air density at current i [kg/m3]
+   real(r8),intent(in)  :: zmagl(pver)          ! height above surface [m]
+   real(r8),intent(in)  :: dz                   ! layer thickness [m]
+   real(r8),intent(in)  :: mu_i(pverp)          ! mu at current i (note pverp dimension, see ma_convproc_tend) [mb/s]
+   real(r8),intent(in)  :: eudp(pver)           ! eu(i,k)*dp(i,k) at current i [mb/s]
+   real(r8),intent(in)  :: const(pcnst_extd,pver)   ! gathered tracer array [kg/kg]
+
+   real(r8), intent(in) :: t(pcols,pver)        ! Temperature [K]
```
Comment (singhbalwinder) [inline components/eam/src/chemistry/modal_aero/modal_aero_convproc.F90:1508]: You will probably revisit this, so, please change these one-letter variables the next time.

File: components/eam/src/chemistry/modal_aero/modal_aero_convproc.F90 [hunk 2]
```diff
@@ -1786,12 +1468,556 @@ subroutine ma_convproc_tend(                                           &
...
+   real(r8)     :: dt_u(pver)           ! lagrangian transport time in the updraft [s]
+   real(r8)     :: mu_p_eudp(pver)      ! = mu_i(kp1) + eudp(k) [mb/s]
+   real(r8)     :: cldfrac_i(pver)      ! cldfrac at current icol (with adjustments) [fraction]
+   real(r8)     :: wup(pver)            ! mean updraft vertical velocity at current level updraft [m/s]
+   real(r8)     :: f_ent                ! fraction of updraft massflux that was entrained across this layer == eudp/mu_p_eudp [fraction]
+
+    ! initiate variables
+    kactcnt = 0 ; kactfirst = 1
+    wup(:) = 0.0
+    clw_cut = 1.0e-6    ! Skip levels where icwmr(icol,k) <= clw_cut (=1.0e-6) to
```
Comment (singhbalwinder) [inline components/eam/src/chemistry/modal_aero/modal_aero_convproc.F90]: Declare this as parameter, since it seems to be a constant

File: components/eam/src/chemistry/modal_aero/modal_aero_convproc.F90 [hunk 3]
```diff
@@ -1786,12 +1468,556 @@ subroutine ma_convproc_tend(                                           &
...
+                        ! ZM module
+    dconudt_wetdep(:,:) = 0.0
+    dconudt_activa(:,:) = 0.0
+
+    ! cldfrac = conv cloud fractional area.  This could represent anvil cirrus area,
+    ! and may not useful for aqueous chem and wet removal calculations
+    ! adjust to remove zero clouds
+    do kk = kbot, ktop, -1
+        cldfrac_i(kk) = cldfrac(icol,kk)
+        cldfrac_i(kk) = max( cldfrac_i(kk), 0.005_r8 )
```
Comment (singhbalwinder) [inline components/eam/src/chemistry/modal_aero/modal_aero_convproc.F90]: May be store 0.005_r8 in a parameter variable if it is some sort of cutoff

File: components/eam/src/chemistry/modal_aero/modal_aero_convproc.F90 [hunk 4]
```diff
@@ -1786,12 +1468,556 @@ subroutine ma_convproc_tend(                                           &
...
+    do kk = kbot, ktop, -1
+        kp1 = kk+1
+
+        !Initialized so that it has a value if the following "if" check yeilds
+        !.false.
+        fa_u(kk) = 0.0_r8
+
+        ! mu_p_eudp(k) = updraft massflux at k, without detrainment between kp1,k
+        mu_p_eudp(kk) = mu_i(kp1) + eudp(kk)
+        if (mu_p_eudp(kk) > mbsth) then
```
Comment (singhbalwinder) [inline components/eam/src/chemistry/modal_aero/modal_aero_convproc.F90:1552]: `mbsth` also seems like a constant.

File: components/eam/src/chemistry/modal_aero/modal_aero_convproc.F90 [hunk 5]
```diff
@@ -1786,12 +1468,556 @@ subroutine ma_convproc_tend(                                           &
...
+                      / (rhoair_i(kk) * (cldfrac_i(kk)*0.5_r8))
+       wup(kk) = max( 0.1_r8, wup(kk) )
+
+! deep - the above method overestimates updraft area and underestimate wup
+!    the following is based lemone and zipser (j atmos sci, 1980, p. 2455)
+!    peak updraft (= 4 m/s) is sort of a "grand median" from their GATE data
+!       and Thunderstorm Project data which they also show
+!    the vertical profile shape is a crude fit to their median updraft profile
+   else
+       zkm = zmagl(kk)*1.0e-3
```
Comment (singhbalwinder) [inline components/eam/src/chemistry/modal_aero/modal_aero_convproc.F90:1637]: We should eventually factor out unit conversions. For example, here we can declare`m_to_km = 1.0e-3` and use it for unit conversion.

File: components/eam/src/chemistry/modal_aero/modal_aero_convproc.F90 [hunk 6]
```diff
@@ -1786,12 +1468,556 @@ subroutine ma_convproc_tend(                                           &
...
+!    when kactcnt>1
+!       apply secondary activatation to the entire updraft
+!       do this for all levels above cloud base (even if completely glaciated)
+!          (this is something for sensitivity testing)
+!-----------------------------------------------------------------------
+   use ppgrid, only: pver, pverp, pcols
+   use constituents, only: pcnst
+
+   ! cloudborne aerosol, so the arrays are dimensioned with pcnst_extd = pcnst*2
+   integer,  parameter  :: pcnst_extd = pcnst*2
```
Comment (singhbalwinder) [inline components/eam/src/chemistry/modal_aero/modal_aero_convproc.F90:1668]: We can leave it as is now but `pcnst_extd` logic can be hard to follow. If it is easy, I would like to see if we can explicitly use `pcnst` and treat both interstitial and cloud borne separately.

File: components/eam/src/chemistry/modal_aero/modal_aero_convproc.F90 [hunk 7]
```diff
@@ -1786,12 +1468,556 @@ subroutine ma_convproc_tend(                                           &
...
+
+   logical      :: do_act_this_lev      ! flag for doing activation at current level
+   integer      :: kp1                  ! kk + 1
+
+   ! initiate variables
+   do_act_this_lev = .false.
+   kp1 = kk+1
+
+   if (kactcnt <= 0) then
+       if (icwmr(icol,kk) > clw_cut) then
```
Comment (singhbalwinder) [inline components/eam/src/chemistry/modal_aero/modal_aero_convproc.F90:1691]: Instead of passing `clw_cut`, we can declare it as a parameter at the top of this module. If the variable is a parameter, it is perfectly fine to declare it at the top and use it.

File: components/eam/src/chemistry/modal_aero/modal_aero_convproc.F90 [hunk 8]
```diff
@@ -1786,12 +1468,556 @@ subroutine ma_convproc_tend(                                           &
...
+            if (iflux_method /= 2) then
+               fluxin  =     mu_i(kp1)*conu(icnst,kp1)                          &
+                           + mu_i(kk )*min(chat(icnst,kk ),const(icnst,km1x))   &
+                         - ( md_i(kk )*cond(icnst,kk)                           &
+                           + md_i(kp1)*min(chat(icnst,kp1),const(icnst,kp1x)) )
+               fluxout =     mu_i(kk )*conu(icnst,kk)                           &
+                           + mu_i(kp1)*min(chat(icnst,kp1),const(icnst,kk ))    &
+                         - ( md_i(kp1)*cond(icnst,kp1)                          &
+                           + md_i(kk )*min(chat(icnst,kk ),const(icnst,kk )) )
+            else
```
Comment (singhbalwinder) [inline components/eam/src/chemistry/modal_aero/modal_aero_convproc.F90:1927]: From the code coverage, it looks like the else block is never invoked. Would you please check? Thanks!

**Review by singhbalwinder — APPROVED** [review]
This looks good to me. Thanks for your hard work on this! I have suggested a few changes. You are still working on it, so it is up to you if you would like to address it here or in your next PR.

File: components/eam/src/chemistry/modal_aero/modal_aero_convproc.F90 (hunk 1 above)
Comment (tangshuaiqi) [reply components/eam/src/chemistry/modal_aero/modal_aero_convproc.F90:1508]: will revisit in the next PR

File: components/eam/src/chemistry/modal_aero/modal_aero_convproc.F90 (hunk 2 above)
Comment (tangshuaiqi) [reply components/eam/src/chemistry/modal_aero/modal_aero_convproc.F90]: will do in this PR

File: components/eam/src/chemistry/modal_aero/modal_aero_convproc.F90 (hunk 3 above)
Comment (tangshuaiqi) [reply components/eam/src/chemistry/modal_aero/modal_aero_convproc.F90]: will do in this PR

File: components/eam/src/chemistry/modal_aero/modal_aero_convproc.F90 (hunk 4 above)
Comment (tangshuaiqi) [reply components/eam/src/chemistry/modal_aero/modal_aero_convproc.F90:1552]: will check it. if it is constant, will change it to parameter

File: components/eam/src/chemistry/modal_aero/modal_aero_convproc.F90 (hunk 7 above)
Comment (tangshuaiqi) [reply components/eam/src/chemistry/modal_aero/modal_aero_convproc.F90:1691]: will do in this PR

File: components/eam/src/chemistry/modal_aero/modal_aero_convproc.F90 (hunk 8 above)
Comment (tangshuaiqi) [reply components/eam/src/chemistry/modal_aero/modal_aero_convproc.F90:1927]: I will check it and remove if not invoked.

**Review by mingxuanwupnnl — APPROVED** [review]

**Comment by tangshuaiqi** [comment]
All tests PASS. merge into refactor-maint-2.0
